src/vm/analysis/type_checker/mod.rs

```diff
@@ -224,15 +226,67 @@ impl FunctionType {
                 analysis_typecheck_cost(accounting, &TypeSignature::IntType, first)?;
                 analysis_typecheck_cost(accounting, &TypeSignature::IntType, second)?;
 
-                if first != &TypeSignature::IntType && first != &TypeSignature::UIntType {
+                // Note: Clarity2 expanded the comparable types to include ASCII, UTF8 and Buffer.
+                // Int and UInt have been present since Clarity1.
+                let is_clarity2: bool = clarity_version == ClarityVersion::Clarity2;
+                // Step 1: Check the first argument on its own, to see that the first argument
+                // has a supported type according to this ClarityVersion.
+                let first_ok = match first {
```

Comment:
Given the number of matching criteria, I wonder if it makes more sense to introduce a `fn is_comparable(clarity_version: ClarityVersion) -> bool` method in `Value` that simply tells you whether or not arithmetic comparison operators can apply to `Value`.  Then, you'd only need to check that `first` and `second` are comparable in this version of Clarity, and then check that they have the same type.